enterprise_gateway/services/processproxies/yarn.py

```diff
@@ -94,6 +102,89 @@ def launch_process(self, kernel_cmd, **kwargs):
+        when and how checks are applied.
+
+        Confirms if the yarn queue has capacity to handle the resource requests that
+        will be sent to it.
+
+        First check ensures the driver and executor memory request falls within
+        the container size of yarn configuration. This check requires executor and
+        driver memory to be available in the env.
+
+        Second,Current version of check, takes into consideration node label partitioning
+        on given queues. Provided the queue name and node label this checks if
+        the given partition has capacity available for kernel startup.
+
+        All Checks are optional. If we have KERNEL_EXECUTOR_MEMORY and KERNEL_DRIVER_MEMORY
+        specified, first check is performed.
+
+        If we have KERNEL_QUEUE and KERNEL_NODE_LABEL specified, second check is performed.
+
+        Proper error messages are sent back for user experience
+        :param kwargs:
+        :return:
+        """
+        env_dict = kwargs.get('env', {})
+
+        executor_memory = int(env_dict.get('KERNEL_EXECUTOR_MEMORY', 0))
+        driver_memory = int(env_dict.get('KERNEL_DRIVER_MEMORY', 0))
+
+        if executor_memory * driver_memory > 0:
+            container_memory = self.resource_mgr.cluster_node_container_memory()
+            if max(executor_memory, driver_memory) > container_memory:
+                self.log_and_raise(http_status_code=500,
+                                   reason="Container Memory not sufficient for a executor/driver allocation")
+
+        candidate_queue_name = (env_dict.get('KERNEL_QUEUE', None))
+        node_label = env_dict.get('KERNEL_NODE_LABEL', None)
+        partition_availability_threshold = float(env_dict.get('PARTITION_THRESHOLD',95.0))
+
+        if candidate_queue_name is None or node_label is None:
+            return
+
+        # else the resources may or may not be available now. it may be possible that if we wait then the resources
+        # become available. start  a timeout process
+
+        self.start_time = RemoteProcessProxy.get_current_time()
+        self.candidate_queue = self.resource_mgr.cluster_scheduler_queue(candidate_queue_name)
+        self.candidate_partition = self.resource_mgr.cluster_queue_partition(self.candidate_queue, node_label)
+
+        yarn_available = False
+        retry_attempt = 0
+
+        self.log.debug("Yarn Endpoint to check for resources : " + self.yarn_endpoint)
+        self.log.debug("Checking if " + self.candidate_partition + " has used capacity <= " + str(partition_availability_threshold) + "%" )
+        self.log.debug("Retrying for " + str(self.yarn_resource_check_wait_time) + " ms if resources not found ")
+        while not yarn_available:
+            self.handle_yarn_queue_timeout()
+            yarn_available = self.resource_mgr.cluster_scheduler_queue_availability(self.candidate_partition,partition_availability_threshold)
```

Comment:
We should either copy this statement outside the loop so that we don't incur unnecessary sleep calls or switch the two and make the handle-timeout method call conditional.  I prefer the former.  It would be good to produce the 'Retrying for' debug statement ONLY if yarn_available is initially false after the first call.